### ISSUE TITLE
Refreshing QandA maker KB from a SharePoint URL does not work - authentication issue?

### ISSUE THREAD

**Issue by metkhoo** [microsoft/BotFramework-Composer#8957]
## Describe the bug
Cannot extract question answer pairs from Word document held in SharePoint via URL

## Version

V2.1.1
![BotComposerVersion](https://user-images.githubusercontent.com/18167332/152182993-46b53179-2d5f-4399-9ad0-dcb01f04b7b3.png)

## Browser

- [ ] Electron distribution
- [ ] Chrome
- [ ] Safari
- [ ] Firefox
- [x] Edge

## OS
- [ ] macOS
- [x] Windows 11
- [ ] Ubuntu

## To Reproduce

1. Go to SharePoint.  Select a document in a Document Library, Right click, "Copy Link"
2. Open BOT Composer, from main vertical navigator select "Knowlege base". 
3. Next to the bot name, click horizontal ellipses, select "Add QnA Maker knowledge base"
4. Enter name of knowledge base and click "Next"
5. Select "Create new knowledge base from URL" option.  Add Source URL previously copied from SharePoint, leave language as "English (United States)".  Leave multi-turn extraction off.  Click "Next"
6. Get message "Knowlege base is ready" (no errors)

## Expected behavior

Expect to see the question and answer pairs in the Bot Composer.  Instead, see nothing but the name of the local KB which is itself a link to the correct document.

## Screenshots

What I see is No Questions and No Answers

![NoQnA](https://user-images.githubusercontent.com/18167332/152184375-79f7489a-b22c-4b27-88c0-583dd108bf06.png)

Also the "add new question" does not work so it is now not possible to add questions manually to the KB.

## Additional context
It may be a security/authentication issue because if I put the word document onto a SharePoint page and then use the SharePoint page as the URL for the KB then I get the following:

![QnAError2](https://user-images.githubusercontent.com/18167332/152185433-31c1b849-904c-47cf-bcfd-59745ef869c1.png)

BTW:
I am logged in to Azure correctly for the bot
I am logged into SharePoint on a browser correctly to access the page

**Comment by metkhoo** [comment]
BTW.  Is there a way to do a programmatic "refresh" if the source URL's have got new content?  I know if you do a "replace content" again and republish that it works but is there a automated version of doing it?

Thanks

**Comment by anishprasad01** [comment]
1. You can look into the [QnA Maker REST API](https://docs.microsoft.com/en-us/azure/cognitive-services/qnamaker/quickstarts/quickstart-sdk?pivots=rest)

2. Secondly, it does look like you're hitting an authentication page when you create your KB. Could you attempt to download the file from SharePoint and upload it locally?

**Comment by metkhoo** [comment]
@anishprasad01 - I tried that but I have determined it is not that simple....
The BOT Composer (BC) creates its "own" Knowledge Base (KB) in the QnA Maker portal. (https://www.qnamaker.ai). 

You can see this by going to the QnA Maker portal and you will see the "BOT" KB with the name;

{YOUR BOT}(composer){Language}

However, the content appears to be controlled only within BOT Composer.  If you inject additional content into this KB from the QnA Maker portal (say using the QnA Maker REST API), then the running BOT will ignore those additional QA pairs. 

Moreover, the next time you "PUBLISH" the BOT from BC, it **overwrites** any additional content you may have put into the KB using the API with a copy of the "Local" KB.

One workaround would be to use the REST API to add additional content in the Portal Bot owned KB and then from with BC, you REPLACE the local content with the content from its own database before republishing.  Personally I don't like that workaround so another would be to create a SECOND KB with is native to the Portal, You can use the REST API to update that and then you still need to do a "REPLACE and PUBLISH" in BC to get it all working.

I cannot see any way to get the BC BOT to accept new QA pairs other than republishing the "Local" KB.

I do not know what is going in inside the BOT but it seems like the BOT needs to "know" about the QA pairs at the time of publishing

The original question for this post still applies..

Why can't BC replace content from a SharePoint hosted document.  

A supplementary question that may help to answer the first one is

In what security context does the BC BOT work?  It does not seem to be the same as the Azure Logged in user.  If it WAS, then I'm sure it would work.

**Comment by anishprasad01** [comment]
Since Composer is meant to be a bit easier to work with, it has more limitations than a code-first bot would. As far as I can tell, the only way to avoid having a local KB overwrite a remote KB, would be to not have a local KB at all. You could try creating the remote one first, and then just adding the key to your Composer bot.

As far as SharePoint goes, I don't know what else to tell you beyond that access to SharePoint by Composer like this isn't supported, likely due to security concerns with the application impersonating a user.

You could attempt the workaround of creating a cloud KB and I can confirm with the team, but I am fairly certain there aren't any other workaround besides this one and the ones you outlined.

**Comment by anishprasad01** [comment]
I confirmed with the team, and the workaround of exporting from the cloud KB and updating the local Composer files is the way it currently needs to be done.

**Comment by metkhoo** [comment]
Thanks for the confirmation. 
So to summarise:
* A Bot Composer BOT QnA integration is via a LOCAL KB only
* Any updates (i.e adding addtiional QnA pairs) at the source QnA URLs will not be available to the published BOT until they are firstly added to the LOCAL Bot Composer KB and then only after the BOT is REPUBLISHED.

Did I get that right?

Thanks again

**Comment by anishprasad01** [comment]
Correct. As per current functionality, there is no way to prevent Composer from updating the Cloud KB with its own local KB when you publish a new version. Any new pairs you add and publish to the production slot in QnA should be picked up by the currently running bot, but as soon as you publish the bot they will be overwritten.

**Comment by metkhoo** [comment]
Thanks for confirmation.  But strangely, I found that adding new pairs to the production slot in QnA maker (composer) KB were **not** picked up by the currently running bot.  Should they have been?

Thanks :D

**Comment by metkhoo** [comment]
How is it even possible to NOT have a local KB and still have QnA maker integration with a Bot Composer BOT?

- If I add an existing KB to the BOT settings then it does not find the QnA Maker KB.  
- If I add "Connect to external resources" in the Unknown intent and then the bot responds with "KnowledgeBaseId".  
- If I hard code the QnA Maker KB into the QnAMaker dialog I get the BOT responding with "Host".  
- If I add a QnA Maker KB to the BOT it always creates a LOCAL KB and I am back where I started when posting this question.

Thanks

**Comment by anishprasad01** [comment]
Sorry, let me be more specific. When you have no QnA added, then you add one, the 4th scenario you specify occurs. However, any pairs in the QnA KB at the time you add it should be added.

You are correct that from this point on, a local KB always exists. The only workaround for this is to export the cloud KB and add it locally before you publish the bot.

**Comment by anishprasad01** [comment]
Since we have identified a potential workaround and confirmed that this is expected behavior, I am going to pass your feedback on to the team and close this issue. 

If you have issues or feature requests in the future, please feel free to raise them in the appropriate repository.
